File: CodeForces/scanners/scanner.java
import java.io.*;
import java.util.*;

class Scanner {
    private BufferedReader br;
    private StringTokenizer st;

    public Scanner() {
        br = new BufferedReader(new InputStreamReader(System.in));
        st = null;
    }

    private void refill() {
        try {
            while (st == null || !st.hasMoreTokens()) {
                String line = br.readLine();
                if (line == null) throw new RuntimeException("End of input");
                st = new StringTokenizer(line);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public int nextInt() {
        refill();
        return Integer.parseInt(st.nextToken());
    }

    public double nextDouble() {
        refill();
        return Double.parseDouble(st.nextToken());
    }

    public long nextLong() {
        refill();
        return Long.parseLong(st.nextToken());
    }

    public String nextWord() {
        refill();
        return st.nextToken();
    }

    public String nextString() {
        refill();
        StringBuilder sb = new StringBuilder();
        while (st.hasMoreTokens()) {
            sb.append(st.nextToken());
            if (st.hasMoreTokens()) sb.append(" ");
        }
        st = null;
        return sb.toString();
    }

    public int[] nextIntArray(boolean sorted) {
        refill();
        List<Integer> arrList = new ArrayList<>();
        while (st.hasMoreTokens()) {
            arrList.add(Integer.parseInt(st.nextToken()));
        }
        st = null;
        int[] arr = arrList.stream().mapToInt(Integer::intValue).toArray();
        if (sorted) Arrays.sort(arr);
        return arr;
    }

    public long[] nextLongArray(boolean sorted) {
        refill();
        List<Long> arrList = new ArrayList<>();
        while (st.hasMoreTokens()) {
            arrList.add(Long.parseLong(st.nextToken()));
        }
        st = null;
        long[] arr = arrList.stream().mapToLong(Long::longValue).toArray();
        if (sorted) Arrays.sort(arr);
        return arr;
    }
}

File: CodeForces/scanners/scanner.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

class Scanner {
    private val reader = BufferedReader(InputStreamReader(System.`in`))
    private var st: StringTokenizer? = null

    private fun refill() {
        while (st == null || !st!!.hasMoreTokens()) {
            val line = reader.readLine() ?: throw RuntimeException("EOF reached")
            st = StringTokenizer(line)
        }
    }

    fun nextInt(): Int {
        refill()
        return st!!.nextToken().toInt()
    }

    fun nextDouble(): Double {
        refill()
        return st!!.nextToken().toDouble()
    }

    fun nextLong(): Long {
        refill()
        return st!!.nextToken().toLong()
    }

    fun nextWord(): String {
        refill()
        return st!!.nextToken()
    }

    fun nextString(): String {
        refill()
        val sb = StringBuilder()
        while (st!!.hasMoreTokens()) {
            sb.append(st!!.nextToken())
            if (st!!.hasMoreTokens()) sb.append(" ")
        }
        st = null
        return sb.toString()
    }

    fun nextIntArray(sorted: Boolean = false): IntArray {
        refill()
        val arr = mutableListOf<Int>()
        while (st!!.hasMoreTokens()) {
            arr.add(st!!.nextToken().toInt())
        }
        st = null
        if (sorted) arr.sort()
        return arr.toIntArray()
    }

    fun nextLongArray(sorted: Boolean = false): LongArray {
        refill()
        val arr = mutableListOf<Long>()
        while (st!!.hasMoreTokens()) {
            arr.add(st!!.nextToken().toLong())
        }
        st = null
        if (sorted) arr.sort()
        return arr.toLongArray()
    }
}
